File: Neetcode/WordSearch.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(res, r, c, v):
            if res == word:
                return True
            elif len(res) > len(word):
                return False
            if 0 <= c-1 < len(board[0]) and 0 <= r < len(board) and board[r][c-1] == word[len(res)] and not visited[r][c-1]:
                visited[r][c-1] = True
                if dfs(res+board[r][c-1], r, c-1, visited):
                    return True
                visited[r][c-1] = False
            if 0 <= c < len(board[0]) and 0 <= r+1 < len(board) and board[r+1][c] == word[len(res)] and not visited[r+1][c]:
                visited[r+1][c] = True
                if dfs(res+board[r+1][c], r+1, c, visited):
                    return True
                visited[r+1][c] = False
            if 0 <= c+1 < len(board[0]) and 0 <= r < len(board) and board[r][c+1] == word[len(res)] and not visited[r][c+1]:
                visited[r][c+1] = True
                if dfs(res+board[r][c+1], r, c+1, visited):
                    return True
                visited[r][c+1] = False
            if 0 <= c < len(board[0]) and 0 <= r-1 < len(board) and board[r-1][c] == word[len(res)] and not visited[r-1][c]:
                visited[r-1][c] = True
                if dfs(res+board[r-1][c], r-1, c, visited):
                    return True
                visited[r-1][c] = False
            return False

        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == word[0]:
                    visited = [[False]*len(board[0]) for _ in range(len(board))]
                    visited[r][c] = True
                    if dfs(board[r][c], r, c, visited):
                        return True
        return False

File: Neetcode/WordSearch.cpp
class Solution {
public:
    bool check(vector<vector<char>> board, string word, int i, int j, vector<vector<bool>>& visited, int w){
        if (w == word.size()-1){
            return true;
        }
        visited[i][j] = true;
        bool next = false;
        if (i >= 0 && i < board.size() && j-1 >= 0 && j-1 < board[0].size() && board[i][j-1] == word[w+1] && !visited[i][j-1]){
            next = next || check(board, word, i, j-1, visited, w+1);
            if (next){
                return next;
            }
        }
        if (i >= 0 && i < board.size() && j+1 >= 0 && j+1 < board[0].size() && board[i][j+1] == word[w+1] && !visited[i][j+1]){
            next = next || check(board, word, i, j+1, visited, w+1);
            if (next){
                return next;
            }
        }
        if (i-1 >= 0 && i-1 < board.size() && j >= 0 && j < board[0].size() && board[i-1][j] == word[w+1] && !visited[i-1][j]){
            next = next || check(board, word, i-1, j, visited, w+1);
            if (next){
                return next;
            }
        }
        if (i+1 >= 0 && i+1 < board.size() && j >= 0 && j < board[0].size() && board[i+1][j] == word[w+1] && !visited[i+1][j]){
            next = next || check(board, word, i+1, j, visited, w+1);
            if (next){
                return next;
            }
        }
        visited[i][j] = false;
        return next;
    }
    bool exist(vector<vector<char>>& board, string word) {
        vector<vector<bool>> visited(board.size(), vector<bool> (board[0].size(), false));
        set<char> boardSet;
        set<char> wordSet(word.begin(), word.end());
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                boardSet.insert(board[i][j]);
            }
        }
        // cout << "BS " << boardSet.size();
        // cout << "\n";
        // cout << "WS " << wordSet.size();
        // cout << "\n";
        // for(auto b:boardSet){
        //     cout << b << " ";
        // }
        // cout << "\n";
        // for(auto w:wordSet){
        //     cout << w << " ";
        // }
        // cout << "\n";
        for(auto w:wordSet){
            if(find(boardSet.begin(), boardSet.end(), w) == boardSet.end()){
                // cout << "didnt match";
                return false;
            }
        }
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                if(board[i][j] == word[0] && check(board, word, i, j, visited, 0)){
                    return true;
                }
            }
        }
        return false;
    }
};
